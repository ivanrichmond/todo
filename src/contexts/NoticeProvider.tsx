import { createContext, ReactNode, useContext, useState } from "react";

// This next line is confusing.  The Notice contructor has a type of type TypeOfNotice.
// The type for the entire Notice itself is NoticeType.  TODO: Redo these type declarations in class?
import Notice, {Message, NoticeType, TypeOfNotice} from '../classes/Notice'

export const NoticeContext = createContext(null);

type NoticeProviderProps = {
    children?: ReactNode
}

export const NoticeProvider = ({ children }: NoticeProviderProps) => {
    const defaultNotice = {message: '', type: '', sendToConsole: () => {}}
    //TODO: EIther use defaultNotice or get rid of it.
    const [notice, setNotice] = useState<NoticeType | null>(null)

    const createNotice = (message: Message, type: TypeOfNotice = 'info') => {
        const newNotice: NoticeType = new Notice(message, type)
        //TODO: Fix this typescript error.
        // @ts-expect-error
        newNotice.sendToConsole()
        setNotice(newNotice)
    }

    const deleteNotice = () => {
        setNotice(defaultNotice)
    }

    const value = {notice, createNotice, deleteNotice}
    //TODO: Fix the following TypeScript error, which occurs without @ts-expect-error.
    // "Type '{ notice: NoticeType | null; createNotice: (message: Message, type?: TypeOfNotice) 
    // => void; deleteNotice: () => void; }' is not assignable to type 'null'."
    // @ts-expect-error
    return <NoticeContext.Provider value={value}>{children}</NoticeContext.Provider>;
};

export const useNotice = () => {
  return useContext(NoticeContext);
};