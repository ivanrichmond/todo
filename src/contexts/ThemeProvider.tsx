// A context for theming light and dark mode.
// TODO: This file is currently just a placeholder.  Build this.