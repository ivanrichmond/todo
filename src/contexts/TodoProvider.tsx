// Context for Todos array: [{task, dueDate, status}, ...]
import { createContext, ReactNode, useState } from 'react'
import Todo, { DueDate, Status, Task, TodoType } from '../classes/Todo'
import todosInitial from '../testData/todos.json'

export type TodoContextType = {
    todos?: TodoType[];
    addTodo?: (task: Task, dueDate: DueDate) => void;
    deleteTodo?: (index: number) => void;
    toggleComplete?: (index: number) => void;
    filter?: (status: Status) => void;
}

// TODO: Fix any.  For some reason, TodoContextType | null throws an error.
export const TodoContext = createContext<any>(null);

type TodoProviderProps = {
    children?: ReactNode
}

export const TodoProvider = ({ children }: TodoProviderProps) => {
    let initialData = todosInitial.map(todo => new Todo(todo.task, todo.dueDate, todo.status))

    const [todos, setTodos] = useState(initialData)
    const [filterStatus, setFilterStatus] = useState('all')
    
    // In the following methods, todos must be changed statelessly by making a copy and setting that.

    // add a new todo.
    const addTodo = (task: Task, dueDate: DueDate) => {
        const newTodo = new Todo(task, dueDate)
        let newTodos = [...todos]
        newTodos.push(newTodo)
        setTodos(newTodos)
    }

    // delete a todo, using index.
    const deleteTodo = (index: number) => {
        let newTodos = [...todos]
        newTodos.splice(index,1)
        setTodos(newTodos)
    }

    // set status to 'complete' on a todo, given index.
    const toggleComplete = (index: number) => {
        let newTodos = [...todos]
        newTodos[index].status = newTodos[index].status === 'complete' ? 'incomplete' : 'complete'
        setTodos(newTodos)
    }

    const values = {todos, addTodo, deleteTodo, toggleComplete, filterStatus, setFilterStatus}
    return <TodoContext.Provider value={values}>{children}</TodoContext.Provider>;
};
