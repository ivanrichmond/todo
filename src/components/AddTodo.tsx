import React, { useContext, useState } from 'react'

import AppForm from '../designSystem/AppForm'
import { TodoContext } from '../contexts/TodoProvider'

const AddToDo = () => {
    // states
    const [task, setTask] = useState('')
    const [dueDate, setDueDate] = useState('')

    // contexts
    const { addTodo } = useContext(TodoContext)

    return(
        <AppForm >
            <AppForm.Group inline>
                <AppForm.Input 
                label='New Task' 
                placeholder='New Task' 
                onChange = {(event) => setTask(event.target.value)}
                />
                <AppForm.Input 
                label='Due Date' 
                placeholder='Due Date' 
                type='date' 
                onChange = {(event) => setDueDate(event.target.value)}
                />
                <AppForm.Button icon='plus' onClick = {() => addTodo(task, dueDate)}/>
            </AppForm.Group>
        </AppForm>
    )
}

export default AddToDo