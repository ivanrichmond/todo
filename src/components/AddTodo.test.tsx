import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer'

import AddTodo, { validateAddTodo } from './AddTodo';
import { NoticeProvider } from '../contexts/NoticeProvider'
import { TodoProvider } from '../contexts/TodoProvider'

describe('AddTodo', () => {
    /**
    * This is an extremely basic smoke test.
    * Passing is not meant to indicate that the component is working,
    * so much as failing indicates that something is seriously wrong with the component.
    */
    test('renders to do', () => {
        render(
            <TodoProvider>
                <NoticeProvider>
                    <AddTodo />
                </NoticeProvider>
            </TodoProvider>
        )
        const testText = screen.getByText(/new task/i)
        expect(testText).toBeInTheDocument()
    })

    test('validateAddTodo with valid parameters returns validate.isValid = true', () => {
        const validation = validateAddTodo('Write more unit tests', '8/6/2024')
        expect(validation.isValid).toBe(true)
    })

    // Just make sure that the component starts out rendered as expected.
    test('AddTodo initial render snapshot', () => {
        const component = renderer.create(
            <TodoProvider>
                <NoticeProvider>
                    <AddTodo />
                </NoticeProvider>
            </TodoProvider>
        );
        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
})

