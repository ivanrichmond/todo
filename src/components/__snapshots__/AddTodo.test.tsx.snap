// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo AddTodo initial render snapshot 1`] = `
<form
  className="ui form AddTodo"
  onSubmit={[Function]}
>
  <div
    className="inline fields"
  >
    <label
      htmlFor="New Task"
    >
      New Task
    </label>
    <div
      className="field"
    >
      <div
        className="ui input"
        dataTestid="new-task"
      >
        <input
          aria-describedby={null}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          placeholder="New Task"
          type="text"
          value=""
        />
      </div>
    </div>
    <label
      htmlFor="Due Date"
    >
      Due Date
    </label>
    <div
      className="field AppDate"
    >
      <div
        className="ui input"
      >
        <input
          aria-describedby={null}
          name="Due Date"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          placeholder="Due Date"
          type="date"
        />
      </div>
    </div>
    <div
      className="field"
    >
      <button
        aria-describedby={null}
        className="ui blue icon button"
        dataTestid="add-button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <i
          aria-hidden="true"
          className="plus icon"
          onClick={[Function]}
        />
      </button>
    </div>
  </div>
</form>
`;
